### PR TITLE
Registrar: Deposit covering max code size

### PR DESCRIPTION
**Pull request by Szegoo**
This PR implements phase 1 of: https://github.com/paritytech/polkadot-sdk/pull/2372#discussion_r1461873933

NOTE: This means that all the current parachains can upgrade their code to the maximum size for free.

### CODE REVIEW

**Comment by Szegoo** [comment]
@eskimor @bkchr PTAL.

**Review by eskimor — APPROVED** [review]
Awesome! Thank you @Szegoo !

File: polkadot/runtime/common/src/paras_registrar/mod.rs [hunk 1]
```diff
@@ -661,7 +667,7 @@ impl<T: Config> Pallet<T> {
 		let per_byte_fee = T::DataDepositPerByte::get();
 		let deposit = T::ParaDeposit::get()
 			.saturating_add(per_byte_fee.saturating_mul((genesis_head.0.len() as u32).into()))
-			.saturating_add(per_byte_fee.saturating_mul((validation_code.0.len() as u32).into()));
+			.saturating_add(per_byte_fee.saturating_mul((config.max_code_size).into()));
```
Comment (bkchr) [inline polkadot/runtime/common/src/paras_registrar/mod.rs:670]:
```suggestion
			.saturating_add(per_byte_fee.saturating_mul(config.max_code_size.into()));
```

**Review by bkchr — APPROVED** [review]

File: polkadot/runtime/common/src/integration_tests.rs [hunk 2]
```diff
@@ -930,8 +930,18 @@ fn basic_swap_works() {
 
 		// Deposit is appropriately taken
 		// ----------------------------------------- para deposit --- crowdloan
-		assert_eq!(Balances::reserved_balance(&account_id(1)), (500 + 10 * 2 * 1) + 100);
-		assert_eq!(Balances::reserved_balance(&account_id(2)), 500 + 20 * 2 * 1);
+		let crowdloan_depoist = 100;
```
Comment (DrW3RK) [inline polkadot/runtime/common/src/integration_tests.rs:933]:
```suggestion
		let crowdloan_deposit = 100;
```

File: polkadot/runtime/common/src/integration_tests.rs [hunk 3]
```diff
@@ -930,8 +930,18 @@ fn basic_swap_works() {
...
+		let para_id_deposit = <Test as paras_registrar::Config>::ParaDeposit::get();
+		let code_deposit = configuration::Pallet::<Test>::config().max_code_size *
+			<Test as paras_registrar::Config>::DataDepositPerByte::get();
+
+		// Para 2000 has a genesis head size of 10.
+		assert_eq!(
+			Balances::reserved_balance(&account_id(1)),
+			crowdloan_depoist + para_id_deposit + code_deposit + 10
```
Comment (DrW3RK) [inline polkadot/runtime/common/src/integration_tests.rs:941]:
```suggestion
			crowdloan_deposit + para_id_deposit + code_deposit + 10
```

File: polkadot/runtime/common/src/integration_tests.rs [hunk 4]
```diff
@@ -973,8 +983,8 @@ fn basic_swap_works() {
 		// Deregister on-demand parachain
 		assert_ok!(Registrar::deregister(para_origin(2000).into(), ParaId::from(2000)));
 		// Correct deposit is unreserved
-		assert_eq!(Balances::reserved_balance(&account_id(1)), 100); // crowdloan deposit left over
-		assert_eq!(Balances::reserved_balance(&account_id(2)), 500 + 20 * 2 * 1);
+		assert_eq!(Balances::reserved_balance(&account_id(1)), crowdloan_depoist);
```
Comment (DrW3RK) [inline polkadot/runtime/common/src/integration_tests.rs:986]:
```suggestion
		assert_eq!(Balances::reserved_balance(&account_id(1)), crowdloan_deposit);
```

**Review by DrW3RK — COMMENTED** [review]
While viewing the code, I noticed a small typo. Good job with identifying this issue and resolving it!

**Review by kianenigma — APPROVED** [review]
I didn't read the whole conversation leading to this change but the PR does what it says it does ✅

**Pull request merged**